### DISCUSSION TITLE
where T : interface

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#3579] [Language Ideas]
Please allow constraints to only allow interfaces.

**Comment by HaloFour** [comment]
What are the use cases for such a change?

**Comment by TonyValenti** [comment]
There are libraries that I work with that accept only interfaces as type parameters.  If I provide a class object, I get a runtime error.  I want to make this a compile time error.

**Comment by CyrusNajmabadi** [comment]
>  I want to make this a compile time error.

This can be done entirely with an analyzer. Iirc this was actually one of the examples we originally had for when someone would want to use an analyzer.

**Comment by foxesknow** [comment]
> 
> 
> > I want to make this a compile time error.
> 
> This can be done entirely with an analyzer. Iirc this was actually one of the examples we originally had for when someone would want to use an analyzer.

The downside of relying on an analyzer is that the analyzer can be turned off or suppressed. A constraint will always be enforced by the compiler.

**Comment by YairHalberstadt** [comment]
> the analyzer can be turned off or suppressed

And so, if your consumer suppresses the warning he'll get a runtime error. That doesn't sound particularly problematic - c# cannot stop somebody deliberately shooting themselves in the foot.

**Comment by TonyValenti** [comment]
Another downside of an analyzer is that you have to actually make it too.

**Comment by YairHalberstadt** [comment]
Making an analyzer is a lot simpler than adding a language feature by many orders of magnitude

**Comment by foxesknow** [comment]
> 
> 
> > the analyzer can be turned off or suppressed
> 
> And so, if your consumer suppresses the warning he'll get a runtime error. That doesn't sound particularly problematic - c# cannot stop somebody deliberately shooting themselves in the foot.

Or, if you got a compiler error because you had constrained to an interface but someone tried to use a class then you wouldn't get a runtime error.

**Comment by foxesknow** [comment]
> 
> 
> Making an analyzer is a lot simpler than adding a language feature by many orders of magnitude

That's a valid point. However, given the pointless addition of expression-bodied members to the language it would appear that people are prepared to add language features regardless of the complexity.

**Comment by YairHalberstadt** [comment]
And so? You could still get round it using unsafe, or reflection or whatever. And even if it stopped that, somebody can choose to do `((object)null).GetHashCode()` and get a runtime error.

There's an infinite number of ways to get a runtime error in C#. The language has no interest in stopping that. What it does have an interest towards doing is guiding people away from common pits of failure. An analyzer does that suffiently. If you choose to suppress the error, well then it's your own application your putting a bug in.

**Comment by YairHalberstadt** [comment]
> That's a valid point. However, given the pointless addition of expression-bodied members to the language it would appear that people are prepared to add language features regardless of the complexity.

Expression bodied members are a purely syntactic change with low design and implementation cost. They are used by many developers on every single file. They can not be implemented via an analyzer.

Such a constraint as is suggested here requires syntactic and semantic changes. It would require a huge amount of design work (if you don't believe me work on a spec). It may require changes to the runtime to work with reflection. It would require a vast number of tests (I'm implementing extension GetEnumerator, which sounds pretty simple but I've written many hundreds of tests for it). It would be used very rarely, and is perfectly doable with an analyser.

Whether or not you like expression bodied members, the two are simply not comparable.

**Comment by IS4Code** [comment]
I understand constraints in terms of what is allowed to do with the value itself (except for `new()`, which is a hack anyway). What can you do with an interface type that you cannot do with a class type?

**Comment by sgf** [comment]
bad idea.

**Comment by 333fred** [comment]
@sgf I've hidden your comment. Criticism is fine and expected on this repo, but single comments that just say "this is bad" without further explanation are not.

**Comment by sgf** [comment]
> @sgf I've hidden your comment. Criticism is fine and expected on this repo, but single comments that just say "this is bad" without further explanation are not.

its not for this repo,it's for this issue.

> allow constraints to only allow interfaces. 

is not good.

In fact, developers need systematic constraints. We need stronger, simpler and more straightforward where constraints.
Constraints can effectively prevent function designers from doing too many runtime checks. Avoid potential errors during compilation.
But only allowing the interface can not achieve flexible constraints.
For example, developers want to accept only a few runtime types such as int32,int64. We cannot add special interfaces to these types.These types cannot be accurately described by only interface constraints.

**Comment by CyrusNajmabadi** [comment]
> its not for this repo,it's for this issue.

@333fred is saying: the way you presented your criticism of this issue was not acceptable, and so was hidden.  He's saying *inside our repo* we have a code of conduct in terms of how people need to behave and that you were not following those codes.  

You've expanded your crticism here to an acceptable point.  Thanks!

**Comment by sgf** [comment]
> @333fred is saying: the way you presented your criticism of this issue was not acceptable, and so was hidden. He's saying _inside our repo_ we have a code of conduct in terms of how people need to behave and that you were not following those codes.
> 
> You've expanded your crticism here to an acceptable point. Thanks!

Thank you for your explanation, I seem to understand what he meant.

**Comment by juliusfriedman** [comment]
> I understand constraints in terms of what is allowed to do with the value itself (except for `new()`, which is a hack anyway). What can you do with an interface type that you cannot do with a class type?

umm nothing it is class you can't `create` all interfaces are abstract classes at the lowest level special classes.

```c3
using System;
public class C {
    public abstract class T { }
    
    public static void M<T>() where T : C.T 
    {
    }
}
```

Perhaps that could be changed in the future.

```cil
class nested public auto ansi abstract beforefieldinit T
        extends [System.Private.CoreLib]System.Object
    {
        // Methods
        .method family hidebysig specialname rtspecialname 
            instance void .ctor () cil managed 
        {
            // Method begins at RVA 0x2052
            // Code size 7 (0x7)
            .maxstack 8

            IL_0000: ldarg.0
            IL_0001: call instance void [System.Private.CoreLib]System.Object::.ctor()
            IL_0006: ret
        } // end of method T::.ctor

    } // end of class T
```
